No longer allow clipping items by complex clips

mrobinson opened the pull request:
This is part of an API simplification which should hopefully lead to
less code complexity in the future.

Fixes #1742.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/2512)
<!-- Reviewable:end -->

mrobinson commented:
Gecko try job is here: https://github.com/servo/webrender/pull/2512

staktrace commented:
Presumably you meant https://treeherder.mozilla.org/#/jobs?repo=try&revision=ad251afbee824f371097a3ea2c378faa31043be7

bors-servo commented:
:umbrella: The latest upstream changes (presumably #2507) made this pull request unmergeable. Please resolve the merge conflicts.

mrobinson commented:
@staktrace Thanks for provided the correct link. I was in a bit of a rush this morning.

The branch should be rebased and ready to review now.

glennw commented:
Is there a Gecko / Servo patch you could post here for the API update, that we can use while doing try runs in the interim?

mrobinson commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[direct-composition/src/main_windows.rs, line 135 at r1](https://beta.reviewable.io/reviews/servo/webrender/2512#-L7W8NmDNb0E5-9aVA9B:-L7YooLX9Da4DkxSeWNX:b-iikw9w) ([raw file](https://github.com/servo/webrender/blob/1de1c0c8efda98fd3450e74c915ea392a9996eaf/direct-composition/src/main_windows.rs#L135)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

any reason we aren't doing `push_clip_id` with this complex clip?
</blockquote></details>

It looked like the rounded rectangle here wasn't important for the example, but I can add it back. I didn't want to make too many changes here because I don't have a Windows environment ready for testing. I can spend a little time getting the rendered content back to its original state though.

---

*[wrench/src/rawtest.rs, line 606 at r1](https://beta.reviewable.io/reviews/servo/webrender/2512#-L7W9V1NmDp8NvCvB3Qb:-L7Yp22X4Kqisfc4tTAu:b-jk10zs) ([raw file](https://github.com/servo/webrender/blob/1de1c0c8efda98fd3450e74c915ea392a9996eaf/wrench/src/rawtest.rs#L606)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

any reason this is removed?
</blockquote></details>

I added this one to explicitly test items with their own rounded rectangle local clips. Since I removed that feature, I also removed the test for it. Directly below is the same test, but using separate clipping nodes.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/servo/webrender/2512)*
<!-- Sent from Reviewable.io -->

mrobinson commented:
@glennw I don't think there are any code changes necessary for Gecko (it always uses the PrimitiveInfo constructors) and here is a branch with the changes necessary for Servo: https://github.com/mrobinson/servo/tree/changes-for-localclip-changes. I haven't tested this one, but I'm fairly certain there is no change in behavior.

kvark commented:
@mrobinson looks like this is ready to go?

mrobinson commented:
@kvark Yep. Is it okay to land it?

kvark commented:
@mrobinson well, the try looks good, code is approved by me, and Gecko doesn't (seemingly?) appear to need any changes. Shipit!

mrobinson commented:
@kvark Thanks!

@bors-servo r=kvark

bors-servo commented:
:pushpin: Commit d398096 has been approved by `kvark`

<!-- @bors-servo r=kvark d398096d6fe222bbcd51a4c97bd3612ba6d2e128 -->

bors-servo commented:
:hourglass: Testing commit d398096d6fe222bbcd51a4c97bd3612ba6d2e128 with merge 0291c44130b543977f821421a4320f09537059fd...

glennw commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit d398096d6fe222bbcd51a4c97bd3612ba6d2e128 with merge 45a3b7baa3f8efd7c23a35e877ca319a5d3b50dc...

bors-servo commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/servo/webrender/builds/353641716?utm_source=github_status&utm_medium=notification)

glennw commented:
Blocked by #2522.

mrobinson commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit d398096d6fe222bbcd51a4c97bd3612ba6d2e128 with merge e0fbae7eb48ab264d4b81501ededee53e66b4b53...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.8396), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/U3KoaBTpQi2c36Pk1gqT1A), [status-travis](https://travis-ci.org/servo/webrender/builds/353784085?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing e0fbae7eb48ab264d4b81501ededee53e66b4b53 to master...
